superset-frontend/src/explore/components/DatasourcePanel/index.tsx

```diff
@@ -429,24 +444,27 @@ export default function DataSourcePanel({
       columnSlice,
       inputValue,
       lists.columns.length,
-      lists.metrics.length,
+      lists?.metrics?.length,
       metricSlice,
       search,
       showAllColumns,
       showAllMetrics,
+      dataSourceIsQuery,
       shouldForceUpdate,
     ],
   );
 
   return (
     <DatasourceContainer>
-      <SaveDatasetModal
-        visible={showSaveDatasetModal}
-        onHide={() => setShowSaveDatasetModal(false)}
-        buttonTextOnSave={t('Save')}
-        buttonTextOnOverwrite={t('Overwrite')}
-        datasource={datasource}
-      />
+      {dataSourceIsQuery && (
```

Comment:
this is a bit of a nit, but I would consider flipping this to be `if datasource is not a dataset, then show the modal to save dataset`, only because otherwise, there's business logic in the code.. why does a query show this modal, for example? It's just a little more clear perhaps to do it the opposite way.

Reply:
The other reason is for long-term extendability of this feature. I'm cautious of setting up too much logic based on the datasource type, and instead more on the functionality.. utilize duck typing when possible in other words.